### PR TITLE
Added a Service API to fetch a blob of bytes from a URL

### PR DESCRIPTION
**Pull request by christos-h**
## Motivation

We would the Service to download blobs from the internet.

## Proposal

Add a `fetch_url` API which will return a vector of bytes given a URL string. Here are some extra considerations:

1. How can we make this API more general?
    i.   Should we take arbitrary `reqwests`
    ii.  Should we provide our own comprehensive HTTP API?
2. Use `url::Url` instead?

### CODE REVIEW

File: linera-execution/src/lib.rs [hunk 1]
```diff
@@ -414,6 +416,9 @@ pub trait ServiceRuntime: BaseRuntime {
         queried_id: UserApplicationId,
         argument: Vec<u8>,
     ) -> Result<Vec<u8>, ExecutionError>;
+
+    /// Get a blob of bytes from an arbitrary URL.
+    fn get_blob_from_url(&mut self, url: &str) -> Result<Vec<u8>, ExecutionError>;
```
Comment (jvff) [inline linera-execution/src/lib.rs:421]: Optional: I'm wondering if `download_blob` is clearer :thinking:

File: linera-execution/src/lib.rs [hunk 2]
```diff
@@ -414,6 +416,9 @@ pub trait ServiceRuntime: BaseRuntime {
         queried_id: UserApplicationId,
         argument: Vec<u8>,
     ) -> Result<Vec<u8>, ExecutionError>;
+
+    /// Get a blob of bytes from an arbitrary URL.
```
Comment (jvff) [inline linera-execution/src/lib.rs:420]: Nit: conjugation in the documentation

```suggestion
    /// Gets a blob of bytes from an arbitrary URL.
```

File: linera-execution/src/runtime.rs [hunk 3]
```diff
@@ -1223,6 +1223,12 @@ impl ServiceRuntime for ServiceSyncRuntime {
         self.inner().pop_application();
         Ok(response)
     }
+
+    /// Get a blob of bytes from an arbitrary URL.
```
Comment (jvff) [inline linera-execution/src/runtime.rs:1227]: Nit: same here

```suggestion
    /// Gets a blob of bytes from an arbitrary URL.
```

File: linera-execution/src/wasm/system_api.rs [hunk 4]
```diff
@@ -260,6 +260,11 @@ macro_rules! impl_service_system_api {
                 ServiceRuntime::try_query_application(self, application.into(), argument.to_vec())
             }
 
+            /// Get a blob of bytes from an arbitrary URL.
```
Comment (jvff) [inline linera-execution/src/wasm/system_api.rs:263]: Nit: same here

```suggestion
            /// Gets a blob of bytes from an arbitrary URL.
```

**Review by jvff — APPROVED** [review]

File: linera-execution/src/lib.rs [hunk 5]
```diff
@@ -129,6 +129,8 @@ pub enum ExecutionError {
     OwnerIsNone,
     #[error("Application is not authorized to perform system operations on this chain: {0:}")]
     UnauthorizedApplication(UserApplicationId),
+    #[error("Failed to make network reqwest")]
```
Comment (afck) [inline linera-execution/src/lib.rs:132]:
```suggestion
    #[error("Failed to make network request")]
```
Or "request with reqwest"?

**Review by afck — APPROVED** [review]
I wonder whether we should make this optional (maybe even opt-in); I feel like there should be a way to query a service that guarantees that it won't make any network requests. (But that's a separate issue, of course.)

File: linera-execution/src/lib.rs (hunk 1 above)
Comment (christos-h) [reply linera-execution/src/lib.rs:421]: Right I thought about that too - but I wanted to disambiguate this from the upcoming Blob API, so `get_blob_from_url` seemed more clear in that respect.

File: linera-execution/src/lib.rs (hunk 5 above)
Comment (christos-h) [reply linera-execution/src/lib.rs:132]: No you're right, callers don't care about the implementation details of how a request is made.

File: linera-execution/src/lib.rs (hunk 1 above)
Comment (ma2bd) [reply linera-execution/src/lib.rs:421]: Do we need to use "blob" though? (could also be a file or content or bytes) We may use the word blob for a specific meaning in the core protocol soon

**Pull request merged**